lib/utils.js

```diff
@@ -0,0 +1,154 @@
+ * TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+ * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+ * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+'use strict';
+
+const defaultCaveatRegistry = require('./caveatRegistry');
+const vocab = require('./vocab');
+
+const api = {};
+module.exports = api;
+
+const {
+  capabilityDelegationUri,
+  caveatUri,
+  invocationTargetUri,
+  invokerUri,
+  parentCapabilityUri
+} = vocab;
+
+/**
+ * Retrieves a capability's parent's invoker
+ *
+ * @param {Object} capability - the fully expanded JSON-LD document for the
+ *                              object capability
+ * @param {Object} documentLoader - a JSON-LD document loader
+ *
+ */
+api.getParentInvoker = async ({capability, documentLoader}) => {
+  const {document: parentCapability} =
+    await documentLoader(capability[parentCapabilityUri]);
+
+  return parentCapability.invoker;
+};
+
+/**
+ * Plucks the single element of an array that contains a single item.
+ *
+ * @param {Array} array - an array with one item.
+ */
+api.getOneOrDie = array => {
+  if (!Array.isArray(array) || array.length != 1) {
+    throw new Error('Expected an array of size 1');
+  }
+  return array[0];
+};
+
+/**
+ * Retrieves the invoker from a capability.
+ *
+ * @param {Object} capability - the fully expanded JSON-LD document for the
+ *                              object capability
+ */
+api.getInvoker = capability => {
+  const invoker = capability[invokerUri];
+  const id = capability['@id'];
+  if(!invoker || !id) {
+    throw new Error('Invoker not found for capability');
+  }
+
+  return invoker || id;
+};
+
+/**
+ * Verifies the caveats for a given object capability.
+ *
+ * @param {Object} capability - the fully expanded JSON-LD document for the
+ *                              object capability
+ * @param {Object} jsigs - a configured instance of jsonld-signatures
+ * @param {Object} proofPurposeOptions  - a set of options for validating the
+ *                                        proof purpose.
+ */
+api.verifyCapability = async ({capability, jsigs, proofPurposeOptions}) => {
+  // Capability does not point to a parent capability and it was capable of
+  // delegating a capability
+  if(!(parentCapabilityUri in capability) &&
+    capabilityDelegationUri in capability) {
+    const {expectedTarget} = proofPurposeOptions;
+    // target is the root capability unless an invocation target is specified
+    const target = invocationTargetUri in capability ?
+      api.getOneOrDie(capability[invocationTargetUri]) : capability['@id'];
+    // It's the toplevel capability, which means it's valid if the target
+    // matches the expected target
+    return target === expectedTarget;
```

Comment:
I'm not sure why the `target === expectedTarget` got moved to verifyCapability, as opposed to the check on the target happening at invocation.  The target is specified in the invocation... why check on every invocation document?  That's extra unnecessary checks.  Plus we may want to check whether a capability chain is valid without being concerned about the intended invocation target.  I'd put this check back where it was.

Reply:
Oh wait, it won't happen over and over again, I was wrong, since it only happens in the toplevel capability.

Still, it does prevent checking whether a capability *chain* is valid, as opposed to an *invocation*

Reply:
> I'm not sure why the `target === expectedTarget` got moved to verifyCapability, as opposed to the check on the target happening at invocation. The target is specified in the invocation... why check on every invocation document? That's extra unnecessary checks. 


Previous Implementation: [link](https://github.com/digitalbazaar/ocapld.js/blob/60a582ad5473a58a76b8b783ef2887104512091c/lib/CapabilityInvocation.js#63)

Based off of the previous implementation, the verify method for `CapabilityInvocation` retrieved the target by traversing the chain until it reached the root capability and then retrieved the expected target via `proofPurposeOptions`. Then ran a check on line 74, that would cause the overall verification to fail.

Current Implementation:  [link](https://github.com/digitalbazaar/ocapld.js/pull/1/files/78f8e56c2dc143c395c6447da87602649bdba7e5#diff-50e3aa130a4f97a42ee2cf111c7b1d9dR105)

The current implementation follows the same capability chain as the previous implementation and does not check on every invocation document. It checks only when it reaches the root capability, so there are no unneccessary checks. It makes the same single check as before. This decision was made because we were traversing the chain twice: the first time to verify the target was the expected target and the second time was to verify the chain. In the case that you have to go out to the network to retrieve these capabilities, it makes verification twice as long.

> Plus we may want to check whether a capability chain is valid without being concerned about the intended invocation target.

Would it be sufficient to have a more detailed output that signifies that verification failed?
```
{
  valid: false,
  targetMatchesExpected: false,
  validCapabilityChain:  true
}
```

Reply:
I'm realizing why this was moved here was probably that @gannan08 probably didn't want to deal with reasoning about my solution, which involved providing another recursive utility ;)

Reply:
@cwebber Well, I can't lie and say that wasn't apart of it. Ultimately, this decision was made because we were traversing the chain twice: 1) verify the target was the expected target and 2) verify the chain. In the case that you have to go over to the network to retrieve these capabilities, it makes verification twice as long.

Reply:
So there's a reason why we might want them to be separate, though.  Might you want to check if a *capability chain* is structurally valid, one that isn't currently being invoked?

It's a use case we're not performing yet, but one that might be performed in the future.  I intentionally separated the logic for "check that the chain is valid" and "check that the invocation is valid".  This change reverts that feature.

That said, it does remove the ability to have to fetch twice or handle the caching implications to make that efficient.  Dunno what the right call is.

Reply:
If checking to see if a capability chain is structurally valid is a useful thing, we should make that a utility function.